### PR TITLE
Finalize estimator tags and fix sparse test failure

### LINKED ISSUE #82: Finalize estimator tags
## Context

`scikit-learn==1.6.0` overhauled the [estimator tags](https://scikit-learn.org/dev/developers/develop.html#estimator-tags). #79 and #80 refactored our estimators to be compatible with the new system, which raised some good discussion on the accuracy of our current tags

## Current Tags

I was struggling to keep track of which tags are defined and/or inherited, so I made the figure below to show the boolean tags (excludes `estimator_type` and `transformer_tags.preserves_dtype`) for all our estimators (gist [here](https://gist.github.com/aazuspan/c0f2f51fe3f37a6c301437e5ca1dc493) if we need to recreate it).

yellow=`True`, purple=`False`, white=`NaN`

plot_tags

## Tags

We can expand this as we go, but here are the things that I think we've already discussed or that stick out to me. Please correct me anywhere I'm wrong @grovduck.

### `estimator_type`

Not included in the figure above, but our regressors are all listed as `regressor` while our transformers are `None`. This is consistent with `sklearn` transformers, but seems wrong since `transformer` [is listed](https://scikit-learn.org/dev/modules/generated/sklearn.utils.Tags.html) as an option. 

There's probably no benefit to adding this, but I don't think it would hurt either.

### `target_tags.single_output`

See https://github.com/lemma-osu/sknnr/pull/79#discussion_r1882609450. I think where we've settled is that `MSNRegressor` and `CCorATransformer` *should* support single output as 1) it is technically possible, 2) there are valid applications e.g. single-species models, 3) it is supported by `yaImpute`. No change required there.

`GNNRegressor` and `CCATransformer` do *not* support single output, so those should be explicitly set to `False`.

### `target_tags.required`

I believe this should be `True` for CCA and CCorA, based on their strict requirements. Currently it's only `True` for CCA.

### `target_tags.one_d_labels` and `target_tags.two_d_labels`

These are [defined as](https://scikit-learn.org/stable/modules/generated/sklearn.utils.TargetTags.html): 

> Whether the input is a 1D labels (y).

and 

> Whether the input is a 2D labels (y).

I find those a little confusing (grammar aside), but I would have assumed that they correspond with `input_tags.one_d_array` and `input_tags.two_d_array` and represent support for `y` array shapes, since "label" is sometimes used interchangeably to refer to dependent/target variables. However, the only place I saw this set to `True` in the `sklearn` code base was in `LabelEncoder` and `LabelBinarizer`, which makes me think they're using a more specific definition of "label".

### PR DESCRIPTION
**Pull request by aazuspan**
Closes #82 and #83

- CCA now correctly requires positive targets
- CCorA now correctly requires targets
- All estimators (other than `RawKNNRegressor`) now explicitly prohibit sparse data. This is a temporary measure to fix failing checks. A closer investigation of sparse support is needed.
- The addition of `target_tags.required` to CCorA triggered an estimator check `check_requires_y_none` which looks for an expected error message when passed `y=None`. That error is now correctly raised by passing y into `_validate_data`. This removed the need to separately check the y array. Note that we don't mutate `X` with the validation check as this will strip feature names that must be passed to `StandardScalerWithDOF`.

@grovduck, please check and confirm that I covered all the tag changes we discussed, thanks!

The updated boolean tags following these changes:

image

### CODE REVIEW

File: src/sknnr/_base.py [hunk 1]
```diff
@@ -149,6 +149,12 @@ def kneighbors(
             return_dataframe_index=return_dataframe_index,
         )
 
+    def __sklearn_tags__(self):
```
Comment (grovduck) [inline src/sknnr/_base.py:152]: Stupid question, but why did you set `tags.input_tags.sparse` here instead of in `RawKNNRegressor`.  I notice in your matrix that `sparse = True` for `RawKNNRegressor`.  Is that intentional?

**Review by grovduck — APPROVED** [review]
Looks good to me!  One small question, but otherwise looks ready to merge.

File: src/sknnr/_base.py (hunk 1 above)
Comment (aazuspan) [reply src/sknnr/_base.py:152]: My rationale was that the minor differences between `RawKNNRegressor` and `KNeighborsRegressor` shouldn't affect sparse data support, so I was pretty confident inheriting that without any further investigation (beyond passing estimator checks). What do you think? Better to set everything `False` to start with?

Comment (grovduck) [reply src/sknnr/_base.py:152]: That's kind of what I figured.  We might want to slightly edit the top post in this PR to be:

> * All estimators now explicitly prohibit sparse data (other than `RawKNNRegressor)

I'm fine with keeping what you have as-is.

**Comment by grovduck** [comment]
@aazuspan, looks like we can close #83 as well?

**Comment by grovduck** [comment]
@aazuspan, for some reason #83 still looks to be open?  I think I [just learned](https://github.com/nus-cs2103-AY2021S2/forum/issues/293) that you have to use "closes #82, closes #83" and that the "and" doesn't work?  I may be off, though ...

**Comment by aazuspan** [comment]
Thanks for the heads up! I just closed manually, but good to know about linking multiple issues :)
